Core/MIPS/IR/IRCompALU.cpp

```diff
@@ -0,0 +1,381 @@
+	case 0: CompShiftImm(op, IROp::ShlImm, sa); break; //sll
+	case 2: CompShiftImm(op, (rs == 1 ? IROp::RorImm : IROp::ShrImm), sa); break;	//srl
+	case 3: CompShiftImm(op, IROp::SarImm, sa); break; //sra
+	case 4: CompShiftVar(op, IROp::Shl, IROp::ShlImm); break; //sllv
+	case 6: CompShiftVar(op, (fd == 1 ? IROp::Ror : IROp::Shr), (fd == 1 ? IROp::RorImm : IROp::ShrImm)); break; //srlv
+	case 7: CompShiftVar(op, IROp::Sar, IROp::SarImm); break; //srav
+	default:
+		DISABLE;
+		break;
+	}
+}
+
+void IRJit::Comp_Special3(MIPSOpcode op) {
+	CONDITIONAL_DISABLE;
+	MIPSGPReg rs = _RS;
+	MIPSGPReg rt = _RT;
+
+	int pos = _POS;
+	int size = _SIZE + 1;
+	u32 mask = 0xFFFFFFFFUL >> (32 - size);
+
+	// Don't change $zr.
+	if (rt == 0)
+		return;
+
+	switch (op & 0x3f) {
+	case 0x0: // ext
+		if (pos != 0) {
+			ir.Write(IROp::ShrImm, rt, rs, pos);
+			ir.Write(IROp::AndConst, rt, rt, ir.AddConstant(mask));
+		} else {
+			ir.Write(IROp::AndConst, rt, rs, ir.AddConstant(mask));
+		}
+		break;
+
+	case 0x4: //ins
+	{
+		logBlocks = 1;
+		u32 sourcemask = mask >> pos;
+		u32 destmask = ~(sourcemask << pos);
+		ir.Write(IROp::AndConst, IRTEMP_0, rs, ir.AddConstant(sourcemask));
+		if (pos != 0) {
+			ir.Write(IROp::ShlImm, IRTEMP_0, IRTEMP_0, pos);
+		}
+		ir.Write(IROp::AndConst, rt, rt, ir.AddConstant(destmask));
+		ir.Write(IROp::Or, rt, rt, IRTEMP_0);
+	}
+	break;
+	}
+}
+
+
+void IRJit::Comp_Allegrex(MIPSOpcode op) {
+	CONDITIONAL_DISABLE;
+	MIPSGPReg rt = _RT;
+	MIPSGPReg rd = _RD;
+	// Don't change $zr.
+	if (rd == 0)
+		return;
+
+	switch ((op >> 6) & 31) {
+	case 16: // seb	// R(rd) = (u32)(s32)(s8)(u8)R(rt);
+		ir.Write(IROp::Ext8to32, rd, rt);
+		break;
+
+	case 24: // seh
+		ir.Write(IROp::Ext16to32, rd, rt);
+		break;
+
+	case 20: //bitrev
+	default:
+		Comp_Generic(op);
+		return;
+	}
+}
+
+void IRJit::Comp_Allegrex2(MIPSOpcode op) {
+	CONDITIONAL_DISABLE;
+	MIPSGPReg rt = _RT;
+	MIPSGPReg rd = _RD;
+	// Don't change $zr.
+	if (rd == 0)
+		return;
+
+	switch (op & 0x3ff) {
+	case 0xA0: //wsbh
+		ir.Write(IROp::BSwap16, rd, rt);
+		break;
+	case 0xE0: //wsbw
+		ir.Write(IROp::BSwap32, rd, rt);
+		break;
+	default:
+		Comp_Generic(op);
+		break;
+	}
+}
+
+void IRJit::Comp_MulDivType(MIPSOpcode op) {
+	CONDITIONAL_DISABLE;
+	MIPSGPReg rt = _RT;
+	MIPSGPReg rs = _RS;
+	MIPSGPReg rd = _RD;
+
+	switch (op & 63) {
+	case 16: // R(rd) = HI; //mfhi
+		if (rd != MIPS_REG_ZERO) {
+			ir.Write(IROp::MfHi, rd);
+		}
+		break;
+
+	case 17: // HI = R(rs); //mthi
+		ir.Write(IROp::MtHi, 0, rs);
+		break;
+
+	case 18: // R(rd) = LO; break; //mflo
+		if (rd != MIPS_REG_ZERO) {
+			ir.Write(IROp::MfLo, rd);
+		}
+		break;
+
+	case 19: // LO = R(rs); break; //mtlo
+		ir.Write(IROp::MtLo, 0, rs);
+		break;
+
+	case 24: //mult (the most popular one). lo,hi  = signed mul (rs * rt)
+		ir.Write(IROp::Mult, 0, rs, rt);
+		break;
+
+	case 25: //multu (2nd) lo,hi  = unsigned mul (rs * rt)
+		ir.Write(IROp::MultU, 0, rs, rt);
+		break;
+
+	case 26: //div
+		DISABLE;
```

Comment:
Should probably remove these DISABLEs, unless they're not working?  Maybe a comment if so?

-[Unknown]